soc/xtensa/intel_adsp/common/ipc.c

```diff
@@ -205,12 +208,110 @@ static int dt_init(const struct device *dev)
 	return intel_adsp_ipc_init(dev);
 }
 
+#ifdef CONFIG_PM_DEVICE
+
+void intel_adsp_ipc_set_resume_handler(const struct device *dev,
+	intel_adsp_ipc_resume_handler_t fn, void *arg)
+{
+	struct ipc_control_driver_api *api =
+		(struct ipc_control_driver_api *)dev->api;
+	struct intel_adsp_ipc_data *devdata = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&devdata->lock);
+
+	api->resume_fn = fn;
+	api->resume_fn_args = arg;
+
+	k_spin_unlock(&devdata->lock, key);
+}
+
+void intel_adsp_ipc_set_suspend_handler(const struct device *dev,
+	intel_adsp_ipc_suspend_handler_t fn, void *arg)
+{
+	struct ipc_control_driver_api *api =
+		(struct ipc_control_driver_api *)dev->api;
+	struct intel_adsp_ipc_data *devdata = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&devdata->lock);
+
+	api->suspend_fn = fn;
+	api->suspend_fn_args = arg;
+
+	k_spin_unlock(&devdata->lock, key);
+}
+
+/**
+ * @brief Manages IPC driver power state change.
+ *
+ * @param dev IPC device.
+ * @param action Power state to be changed to.
+ * @return int Returns 0 on success or optionaly error code from the
+ * registered ipc_power_control_api callbacks.
+ */
+static int ipc_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	int ret = 0;
+	const struct ipc_control_driver_api *api =
+		(const struct ipc_control_driver_api *)dev->api;
+
+	/* All device pm is handled during resume and suspend */
+	switch (action) {
+	case PM_DEVICE_ACTION_SUSPEND:
+		pm_device_state_lock(dev);
```

Comment:
Why are you locking the device state here and unlocking after calling `api->suspend_fn`. The state just changes if the system is suspended / resumed what only happens if idle thread runs.

Reply:
This is a validation mechanism provided by the Device Power API.

> The state just changes if the system is suspended / resumed what only happens if idle thread runs.

This is an assumption. Today, for MTL, this is done using only this use case ("what only happens if idle thread runs").
Tomorrow, there will be some additional use case, and system will behave in the uncontrolled way.
Device power state should be changed atomically.
That means, only if the transition of the power state had been successful, we change the Device power state.
That is why, we take a `pm_device_state_lock` and release the lock just before successful completion of the transition logic.
If the Device fails to change power state (and assertion is called), its power state should remain unchanged (this sentence looks weird but well...).
Quoting docs of the [pm_device_state_lock](https://docs.zephyrproject.org/apidoc/latest/group__subsys__pm__device.html#gaab27e932950e1063b2f1f4c4e19dbf01):

> This function locks the current device power state. Once locked the device power state will not be changed by system power management or device runtime power management until unlocked.

Reply:
If you really don't want the system doing power management in devices and have to do it completely manually, please enable `CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE` and use device runtime. That is cleanest way to do it.

Reply:
Can you explaing when this function is preempted by the idle thread ? 

This whole logic is not making sense for me, if you manually controlling power management for this device, you should *not* do it using `pm_device_action_run` but through `pm_device_runtime_get/put`. The latter ensure that the system won't interrupt an operation and do the bookkepping of who is using the driver ... Manually invoke `pm_device_action_run` and use system power management is not right. This callback should just care about do the action it was called and not care about locking, state and other things.

Reply:
> If you really don't want the system doing power management in devices and have to do it completely manually, please enable `CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE` and use device runtime. That is cleanest way to do it.

No it is not. I've been there and tried to implement this initially. Also @tmleman initially proposed usage of `CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE`.
* Device Runtime Power API is used for devices that are put to sleep when not in use. This is not suitable for IPC Device since it should be active at all times when SoC is awake.
* We decided to call manually `pm_device_action_run` functions in the notifiers, because mode simple [Central Method](https://docs.zephyrproject.org/2.6.0/reference/power_management/index.html#central-method) does not meet our needs. In this method, calling function `pm_system_suspend()` would result in the putting to sleep add Devices in the system to D3, but:
    1. We can not determine order of this action
    2. We can not provide any error handling inside this operation (assertions, error codes, etc)
    3. `pm_system_suspend` just ignores devices that are busy:
    > If a decision to enter deep sleep is made, the implementation would enter it only after checking if the devices are not in the middle of a hardware transaction that cannot be interrupted.
    
    This is not a right solution in our case - we receive IPC instructing us to enter D3 and we should just enter it.
    We can optionally take some time waiting for some devices to finish hardware operations and exit busy state,
    but calling `pm_system_suspend()` in a loop until it succeeds seems a bad design to me.
    This Central Method is good for small embedded systems not for our firmware (or at least its current form).

Reply:
> Can you explaing when this function is preempted by the idle thread ?
> 
> This whole logic is not making sense for me, if you manually controlling power management for this device, you should _not_ do it using `pm_device_action_run` but through `pm_device_runtime_get/put`. The latter ensure that the system won't interrupt an operation and do the bookkepping of who is using the driver ... Manually invoke `pm_device_action_run` and use system power management is not right. This callback should just care about do the action it was called and not care about locking, state and other things.

This is simply not a true statement. Function [pm_device_state_set](https://docs.zephyrproject.org/2.6.0/reference/power_management/index.html#device-set-power-state) and [pm_device_state_get](https://docs.zephyrproject.org/2.6.0/reference/power_management/index.html#device-get-power-state) are part of the Zephyr Power Manager **public interface** to be used exactly in the way I do and exactly in the scenario I have (when using [Central Method](https://docs.zephyrproject.org/2.6.0/reference/power_management/index.html#central-method) is not enough).
I've explained why we should not use [Device Runtime Power Management](https://docs.zephyrproject.org/2.6.0/reference/power_management/index.html#device-runtime-power-management) for IPC Device in previous comments.

Reply:
@aborisovich

>Device Runtime Power API is used for devices that are put to sleep when not in use. This is not suitable for IPC Device since it should be active at all times when SoC is awake.

Yes,because some drivers do pm_device_runtime_get/put(self) when do an operation. That is not the case of this device. In this device if you use `CONFIG_PM_DEVICE_RUNTIME` the device will *only* be suspended when you call `pm_device_runtime_put` that is done only when you need. That is exactly alternative for the "Central Method".

>We decided to call manually pm_device_action_run functions in the notifiers, because more simple Central Method does not meet our needs. In this method, calling function pm_system_suspend() would result in the putting to sleep add Devices in the system to D3, but:

That is the whole point of device runtime power management. If pm_device_runtime_get/put is reference counted. If you do a `pm_device_runtime_get` the device will be active until you do `pm_device_runtime_put` and there is no other place holding a reference. e.g: You call `pm_device_runtime_get`, another place calls `pm_device_runtime_get` do something and call `pm_device_runtime_put`, the device will sill be active because the referece count is still 1.

> This is not a right solution in our case - we receive IPC instructing us to enter D3 and we should just enter it. We can optionally take some time waiting for some devices to finish hardware operations and exit busy state, but calling pm_system_suspend() in a loop until it succeeds seems a bad design to me. This Central Method is good for small embedded systems not for our firmware (or at least its current form).

Yes, so why you are relying on this method ? You should be using device runtime pm :)

> This is simply not a true statement. Function pm_device_state_set and pm_device_state_get are part of the Zephyr Power Manager public interface to be used exactly in the way I do and exactly in the scenario I have (when using Central Method is not enough). I've explained why we should not use Device Runtime Power Management for IPC Device in previous comments.

Yes, that is true and we probably need to document it better to not recommend using it directly as you are doing. And the problem is that from the OS perspective if applications use it directly is hard to keep the system sane. Your use case (if I am not understanding wrongly) is the exactly use case for device runtime power management.

Reply:
@aborisovich

>Device Runtime Power API is used for devices that are put to sleep when not in use. This is not suitable for IPC Device since it should be active at all times when SoC is awake.

Yes,because some drivers do pm_device_runtime_get/put(self) when do an operation. That is not the case of this device. In this device if you use `CONFIG_PM_DEVICE_RUNTIME` the device will *only* be suspended when you call `pm_device_runtime_put` that is done only when you need. That is exactly alternative for the "Central Method".

>We decided to call manually pm_device_action_run functions in the notifiers, because more simple Central Method does not meet our needs. In this method, calling function pm_system_suspend() would result in the putting to sleep add Devices in the system to D3, but:

That is the whole point of device runtime power management. If pm_device_runtime_get/put is reference counted. If you do a `pm_device_runtime_get` the device will be active until you do `pm_device_runtime_put` and there is no other place holding a reference. e.g: You call `pm_device_runtime_get`, another place calls `pm_device_runtime_get` do something and call `pm_device_runtime_put`, the device will sill be active because the referece count is still 1.

> This is not a right solution in our case - we receive IPC instructing us to enter D3 and we should just enter it. We can optionally take some time waiting for some devices to finish hardware operations and exit busy state, but calling pm_system_suspend() in a loop until it succeeds seems a bad design to me. This Central Method is good for small embedded systems not for our firmware (or at least its current form).

Yes, so why you are relying on this method ? You should be using device runtime pm :)

> This is simply not a true statement. Function pm_device_state_set and pm_device_state_get are part of the Zephyr Power Manager public interface to be used exactly in the way I do and exactly in the scenario I have (when using Central Method is not enough). I've explained why we should not use Device Runtime Power Management for IPC Device in previous comments.

Yes, that is true and we probably need to document it better to not recommend using it directly as you are doing. And the problem is that from the OS perspective if applications use it directly is hard to keep the system sane. Your use case (if I am not understanding wrongly) is the exactly use case for device runtime power management.

Reply:
@aborisovich

>Device Runtime Power API is used for devices that are put to sleep when not in use. This is not suitable for IPC Device since it should be active at all times when SoC is awake.

Yes,because some drivers do pm_device_runtime_get/put(self) when do an operation. That is not the case of this device. In this device if you use `CONFIG_PM_DEVICE_RUNTIME` the device will *only* be suspended when you call `pm_device_runtime_put` that is done only when you need. That is exactly alternative for the "Central Method".

>We decided to call manually pm_device_action_run functions in the notifiers, because more simple Central Method does not meet our needs. In this method, calling function pm_system_suspend() would result in the putting to sleep add Devices in the system to D3, but:

That is the whole point of device runtime power management. If pm_device_runtime_get/put is reference counted. If you do a `pm_device_runtime_get` the device will be active until you do `pm_device_runtime_put` and there is no other place holding a reference. e.g: You call `pm_device_runtime_get`, another place calls `pm_device_runtime_get` do something and call `pm_device_runtime_put`, the device will sill be active because the referece count is still 1.

> This is not a right solution in our case - we receive IPC instructing us to enter D3 and we should just enter it. We can optionally take some time waiting for some devices to finish hardware operations and exit busy state, but calling pm_system_suspend() in a loop until it succeeds seems a bad design to me. This Central Method is good for small embedded systems not for our firmware (or at least its current form).

Yes, so why you are relying on this method ? You should be using device runtime pm :)

> This is simply not a true statement. Function pm_device_state_set and pm_device_state_get are part of the Zephyr Power Manager public interface to be used exactly in the way I do and exactly in the scenario I have (when using Central Method is not enough). I've explained why we should not use Device Runtime Power Management for IPC Device in previous comments.

Yes, that is true and we probably need to document it better to not recommend using it directly as you are doing. And the problem is that from the OS perspective if applications use it directly is hard to keep the system sane. Your use case (if I am not understanding wrongly) is the exactly use case for device runtime power management.

Reply:
> Tomorrow, there will be some additional use case, and system will behave in the uncontrolled way.
Device power state should be changed atomically.

You have to be specific about the use case that you want to cover. Also, what you are doing is not atomic you can be preempted in the middle of this operation and in that case you may be in an inconsistent state if you try to use this device.

THAT assuming that you are suspending / resuming this device out of the idle loop.

My point is, if you are trying to cover that other places in the code will suspend or resume this device, this operation is not being atomic or if you know that suspend/resume will happen only from the idle thread (or using the proper device runtime API) this lock/unlock mechanism is not needed ...

The purpose of this API is to prevent other places in the code of changing the device state and *NOT* to guarantee state transition atomicity.